From 90aa01a6d3b9e41302f0a3e2cfce6028faaa4f02 Mon Sep 17 00:00:00 2001
From: Pranav Vashi <neobuddy89@gmail.com>
Date: Sun, 7 Jan 2024 01:51:03 +0530
Subject: [PATCH 08/30] SystemUI: biometrics: Follow user setting to confirm
 face unlock

* It seems that AOSP code migration has some pending clean up
  which creates issue of not honoring setConfirmationRequired(false)
  for biometric prompt.
* Let's disregard app setting and follow only user setting whether or
  not confirm face unlock prompt is required.

Change-Id: Idf76c0fc325ed57cf10ade128b42570a54c003c6
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
Signed-off-by: Dmitrii <bankersenator@gmail.com>
---
 .../biometrics/data/repository/PromptRepository.kt     | 10 +---------
 .../biometrics/ui/viewmodel/PromptViewModel.kt         |  9 ++-------
 2 files changed, 3 insertions(+), 16 deletions(-)

diff --git a/packages/SystemUI/src/com/android/systemui/biometrics/data/repository/PromptRepository.kt b/packages/SystemUI/src/com/android/systemui/biometrics/data/repository/PromptRepository.kt
index adc276e4e44e..0c51d1bde4ef 100644
--- a/packages/SystemUI/src/com/android/systemui/biometrics/data/repository/PromptRepository.kt
+++ b/packages/SystemUI/src/com/android/systemui/biometrics/data/repository/PromptRepository.kt
@@ -149,17 +149,9 @@ constructor(
 
     private val _faceSettings =
         _userId.map { id -> faceSettings.forUser(id) }.distinctUntilChanged()
-    private val _faceSettingAlwaysRequireConfirmation =
-        _faceSettings.flatMapLatest { it.alwaysRequireConfirmationInApps }.distinctUntilChanged()
 
-    private val _isConfirmationRequired = _promptInfo.map { it?.isConfirmationRequested ?: false }
     override val isConfirmationRequired =
-        combine(_isConfirmationRequired, _faceSettingAlwaysRequireConfirmation) {
-                appRequiresConfirmation,
-                forceRequireConfirmation ->
-                forceRequireConfirmation || appRequiresConfirmation
-            }
-            .distinctUntilChanged()
+        _faceSettings.flatMapLatest { it.alwaysRequireConfirmationInApps }.distinctUntilChanged()
 
     override fun setPrompt(
         promptInfo: PromptInfo,
diff --git a/packages/SystemUI/src/com/android/systemui/biometrics/ui/viewmodel/PromptViewModel.kt b/packages/SystemUI/src/com/android/systemui/biometrics/ui/viewmodel/PromptViewModel.kt
index 72ad37583be0..4502d781a4b2 100644
--- a/packages/SystemUI/src/com/android/systemui/biometrics/ui/viewmodel/PromptViewModel.kt
+++ b/packages/SystemUI/src/com/android/systemui/biometrics/ui/viewmodel/PromptViewModel.kt
@@ -308,10 +308,7 @@ constructor(
      * If the API caller or the user's personal preferences require explicit confirmation after
      * successful authentication. Confirmation always required when in explicit flow.
      */
-    val isConfirmationRequired: Flow<Boolean> =
-        combine(_isOverlayTouched, size) { isOverlayTouched, size ->
-            !isOverlayTouched && size.isNotSmall
-        }
+    val isConfirmationRequired: Flow<Boolean> = promptSelectorInteractor.isConfirmationRequired
 
     /**
      * When fingerprint and face modalities are enrolled, indicates whether only face auth has
@@ -853,11 +850,9 @@ constructor(
     }
 
     private suspend fun needsExplicitConfirmation(modality: BiometricModality): Boolean {
-        val confirmationRequired = isConfirmationRequired.first()
-
         // Only worry about confirmationRequired if face was used to unlock
         if (modality == BiometricModality.Face) {
-            return confirmationRequired
+            return isConfirmationRequired.first()
         }
         // fingerprint only never requires confirmation
         return false
-- 
2.55.0

